### DISCUSSION TITLE
How can I make MLFlow Python Client with Authentication info?

### DISCUSSION THREAD

**Discussion by hhhyunwoo** [mlflow/mlflow#12881] [Q&A]
Hi folks,

We are currently utilizing the authentication feature provided by MLflow. 
According to the official documentation, it appears that the recommended method for storing the Username and Password is by using OS environment variables. However, this approach seems quite risky.

Instead, I would like to incorporate authentication information directly when creating the Client for an MLflow Tracking Server, as shown in the example below:

```
MlflowClient(
    tracking_uri=self.tracking_uri,
    registry_uri=self.tracking_uri,
    authentication_info=auth
)
```

Is this method possible? If so, could you provide guidance or examples on how to implement this securely?

Thank you!

**Comment by B-Step62** [comment]
> According to the official documentation, it appears that the recommended method for storing the Username and Password is by using OS environment variables. However, this approach seems quite risky.

Unfortunately, MLflow client only supports passing auth info via environment variables or a file. Adding this would be possible but requires some complexity to pass around the auth info across different parts of the code base. As not being an expert in the domain, I'd like to understand the thread model a bit more. Env vars are typically "good-enough" solution for client side, it's only be visible from the user running the process or the root. Passing it directly to the client may reduce some risks but I'm not sure if it protects us when that permission boundary is compromised. If there is a major security hole in current approach, please submit one via https://huntr.com/, the platform we use for managing vulnerabilities.

**Reply by JulianJvn** [reply]
IMO, it's less a question of security (as you said, env. variables are pretty secure), but more one of maintainability and convenience. While the OP body frames it as a security thing, the OP title is independent of that.

The environment variables of a process (`os.environ`) are essentially global state/variables. It's usually bad design for programs to extensively use global state. In this case, I can see the following downsides:

- It makes data flow harder to understand since it introduces a side channel of input data for the program (as opposed to the "main channel" of direct program/function arguments).
- It's unexpected for library users if most things can be set via function arguments, but a specific thing can't.
- It makes parallelism harder since the global state can only have one value at a time.
- It facilitates race conditions if at least one thread reads global state and at least one changes it.

Environment variables are perfectly fine to configure end-user CLI programs (for credentials, it's the preferred option). There, they are not global state, but just another form of supplying program input, akin to CLI options.

However, libraries on the other hand should be reusable as much as possible, which usually means being able to configure everything via function arguments. Environment variables can still be used as a fallback.

The problem with the MLflow Python client using this approach is that this client is and can be used within other libraries, which might be used in contexts where there are multiple users, or where credentials are loaded dynamically after a program has been started, which makes it necessary to set environment variables of the running process. And exactly this causes the mentioned downsides.

I recognize the complexity of changing it, and while there is no reason to change it from a security POV, there are good reasons to do it from other POVs.

**Comment by Labeeb2339** [comment]
As of current MLflow master, `MlflowClient` still has no `authentication_info` argument; its constructor accepts tracking, registry, and workspace-store URIs only. For basic auth, use the supported credentials file instead:

```ini
[mlflow]
mlflow_tracking_username = username
mlflow_tracking_password = password
```

Save it as `~/.mlflow/credentials` with restrictive filesystem permissions. The Python client reads this file, and `MLFLOW_TRACKING_USERNAME/PASSWORD` override it. For bearer auth, use `MLFLOW_TRACKING_TOKEN`; `MLFLOW_TRACKING_AUTH` selects a registered request-auth provider. Direct per-instance credential injection remains unsupported, so avoid mutating process-wide environment variables concurrently for multiple identities. See the [current client signature](https://github.com/mlflow/mlflow/blob/master/mlflow/tracking/client.py#L234-L254), [credential resolution](https://github.com/mlflow/mlflow/blob/master/mlflow/utils/credentials.py#L23-L73), and [credentials-file guidance](https://mlflow.org/docs/latest/self-hosting/security/basic-http-auth/#using-credentials-file).
